From e8441024fcd74a77c5818ae5b174004a1e1d6bbe Mon Sep 17 00:00:00 2001
From: Christophe Guibout <christophe.guibout@st.com>
Date: Fri, 21 Jun 2024 12:54:45 +0200
Subject: [PATCH 666/841] fix(st-ddr): improve DLL clock management in self
 refresh sequence

- Switch off DDRPHY clock before disabling the DLLs
	and after enabling GSKP (glitch skipper)
- Keep GSKP enabled during CStop

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 36cb36703a9a4c6de5ef6ab18406ef801f94f951 fix(st-ddr): improve DLL clock management in self refresh sequence]
Change-Id: I3eea895f00c810649b428611ffcb4498b06ac826
Signed-off-by: Christophe Guibout <christophe.guibout@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/396624
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/ddr/stm32mp1_ddr_helpers.c     | 47 +++++++++++++----------
 drivers/st/ddr/stm32mp1_ram.c             |  3 +-
 include/drivers/st/stm32mp1_ddr_helpers.h |  2 +-
 3 files changed, 29 insertions(+), 23 deletions(-)

diff --git a/drivers/st/ddr/stm32mp1_ddr_helpers.c b/drivers/st/ddr/stm32mp1_ddr_helpers.c
index 5adf38f0f..bf09d8cb5 100644
--- a/drivers/st/ddr/stm32mp1_ddr_helpers.c
+++ b/drivers/st/ddr/stm32mp1_ddr_helpers.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2017-2024, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -16,19 +16,21 @@
 
 #include <platform_def.h>
 
-void ddr_enable_clock(void)
+void ddr_enable_clock(bool ddrphycen)
 {
-	stm32mp1_clk_rcc_regs_lock();
+	uint32_t rcc_ddritfcr = RCC_DDRITFCR_DDRC1EN | RCC_DDRITFCR_DDRPHYCAPBEN |
+				RCC_DDRITFCR_DDRCAPBEN;
 
-	mmio_setbits_32(stm32mp_rcc_base() + RCC_DDRITFCR,
-			RCC_DDRITFCR_DDRC1EN |
 #if STM32MP_DDR_DUAL_AXI_PORT
-			RCC_DDRITFCR_DDRC2EN |
+	rcc_ddritfcr |= RCC_DDRITFCR_DDRC2EN;
 #endif
-			RCC_DDRITFCR_DDRPHYCEN |
-			RCC_DDRITFCR_DDRPHYCAPBEN |
-			RCC_DDRITFCR_DDRCAPBEN);
 
+	if (ddrphycen) {
+		rcc_ddritfcr |= RCC_DDRITFCR_DDRPHYCEN;
+	}
+
+	stm32mp1_clk_rcc_regs_lock();
+	mmio_setbits_32(stm32mp_rcc_base() + RCC_DDRITFCR, rcc_ddritfcr);
 	stm32mp1_clk_rcc_regs_unlock();
 }
 
@@ -103,9 +105,14 @@ static int ddr_sw_self_refresh_in(void)
 	mmio_setbits_32(pwr_base + PWR_CR3, PWR_CR3_DDRRETEN);
 	stm32mp_pwr_regs_unlock();
 
-	/* Switch controller clocks (uMCTL2/PUBL) to DLL ref clock */
 	stm32mp1_clk_rcc_regs_lock();
+
+	/* Switch controller clocks (uMCTL2/PUBL) to DLL ref clock */
 	mmio_setbits_32(rcc_base + RCC_DDRITFCR, RCC_DDRITFCR_GSKPCTRL);
+
+	/* Deactivate DDRPHY clock */
+	mmio_clrbits_32(rcc_base + RCC_DDRITFCR, RCC_DDRITFCR_DDRPHYCEN);
+
 	stm32mp1_clk_rcc_regs_unlock();
 
 	/* Disable all DLLs: GLITCH window */
@@ -128,10 +135,7 @@ static int ddr_sw_self_refresh_in(void)
 
 	stm32mp1_clk_rcc_regs_lock();
 
-	/* Switch controller clocks (uMCTL2/PUBL) to DLL output clock */
-	mmio_clrbits_32(rcc_base + RCC_DDRITFCR, RCC_DDRITFCR_GSKPCTRL);
-
-	/* Deactivate all DDR clocks */
+	/* Deactivate all other DDR clocks */
 	mmio_clrbits_32(rcc_base + RCC_DDRITFCR,
 			RCC_DDRITFCR_DDRC1EN |
 #if STM32MP_DDR_DUAL_AXI_PORT
@@ -162,8 +166,8 @@ int ddr_sw_self_refresh_exit(void)
 	uintptr_t ddrctrl_base = stm32mp_ddrctrl_base();
 	uintptr_t ddrphyc_base = stm32mp_ddrphyc_base();
 
-	/* Enable all clocks */
-	ddr_enable_clock();
+	/* Enable clocks except DDRPHY */
+	ddr_enable_clock(false);
 
 	/*
 	 * Manage quasi-dynamic registers modification
@@ -177,11 +181,6 @@ int ddr_sw_self_refresh_exit(void)
 
 	stm32mp_ddr_unset_qd3_update_conditions((struct stm32mp_ddrctl *)ddrctrl_base);
 
-	/* Switch controller clocks (uMCTL2/PUBL) to DLL ref clock */
-	stm32mp1_clk_rcc_regs_lock();
-	mmio_setbits_32(rcc_base + RCC_DDRITFCR, RCC_DDRITFCR_GSKPCTRL);
-	stm32mp1_clk_rcc_regs_unlock();
-
 	/* Enable all DLLs: GLITCH window */
 	mmio_clrbits_32(ddrphyc_base + DDRPHYC_ACDLLCR,
 			DDRPHYC_ACDLLCR_DLLDIS);
@@ -213,6 +212,12 @@ int ddr_sw_self_refresh_exit(void)
 
 	udelay(DDR_DELAY_10US);
 
+	/* Enable DDRPHY clock */
+	stm32mp1_clk_rcc_regs_lock();
+	mmio_setbits_32(stm32mp_rcc_base() + RCC_DDRITFCR, RCC_DDRITFCR_DDRPHYCEN);
+	stm32mp1_clk_rcc_regs_unlock();
+
+	/* Release reset */
 	mmio_setbits_32(ddrphyc_base + DDRPHYC_ACDLLCR,
 			DDRPHYC_ACDLLCR_DLLSRST);
 
diff --git a/drivers/st/ddr/stm32mp1_ram.c b/drivers/st/ddr/stm32mp1_ram.c
index c04f55f7d..3e9571217 100644
--- a/drivers/st/ddr/stm32mp1_ram.c
+++ b/drivers/st/ddr/stm32mp1_ram.c
@@ -28,7 +28,8 @@ int stm32mp1_ddr_clk_enable(struct stm32mp_ddr_priv *priv, uint32_t mem_speed)
 {
 	unsigned long ddrphy_clk, ddr_clk, mem_speed_hz;
 
-	ddr_enable_clock();
+	/* Enable all clocks, including DDRPHY */
+	ddr_enable_clock(true);
 
 	ddrphy_clk = clk_get_rate(DDRPHYC);
 
diff --git a/include/drivers/st/stm32mp1_ddr_helpers.h b/include/drivers/st/stm32mp1_ddr_helpers.h
index f8c586958..0f44453b5 100644
--- a/include/drivers/st/stm32mp1_ddr_helpers.h
+++ b/include/drivers/st/stm32mp1_ddr_helpers.h
@@ -16,7 +16,7 @@ enum stm32mp1_ddr_sr_mode {
 	DDR_ASR_MODE,
 };
 
-void ddr_enable_clock(void);
+void ddr_enable_clock(bool ddrphycen);
 int ddr_sw_self_refresh_exit(void);
 uint32_t ddr_get_io_calibration_val(void);
 int ddr_standby_sr_entry(void);
-- 
2.47.2

